File: leetcode/lc-us/cpp/2516.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int takeCharacters(string s, int k) {
        int n = (int) s.size();
        int cnt[3] = {};
        for (char c: s) {
            cnt[c - 'a']++;
        }
        for (int i: cnt) {
            if (i < k) return -1;
        }
        int res = n + 1;
        for (int r = 0, l = 0; r < n; r++) {
            int idx = s[r] - 'a';
            cnt[idx]--;
            while (cnt[idx] < k) {
                cnt[s[l] - 'a']++;
                l++;
            }
            res = min(res, n - (r - l + 1));
        }
        return res > n ? -1 : res;
    }
};

File: leetcode/lc-us/py/2516.py
from collections import Counter


class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        n = len(s)
        cnt = Counter(s)
        if k and len(cnt) < 3:
            return -1
        for _, freq in cnt.items():
            if freq < k:
                return -1
        res = n + 1
        i, j = 0, 0
        while j < n:
            c = s[j]
            cnt[c] -= 1
            while cnt[c] < k:
                cnt[s[i]] += 1
                i += 1
            res = min(res, n - (j - i + 1))
            j += 1
        return -1 if res > n else res
